packages/common-all/src/notePropDict.ts

```diff
@@ -0,0 +1,134 @@
+import _ from "lodash";
+import {
+  NotePropsDict,
+  NotePropsByFnameDict,
+  DVault,
+  NoteProps,
+  NotePropsFullDict,
+} from "./types";
+import { cleanName, isNotUndefined } from "./utils";
+import { VaultUtils } from "./vault";
+
+/**
+ * Utilities for working with NotePropsFullDict. The reason NotePropsFullDict is not a class is due to needing
+ * to work with primitive objects with redux
+ */
+export class NoteFullDictUtils {
+  /**
+   * Find notes by fname. If vault is provided, filter results so that only notes with matching vault is returned
+   * Return empty array if no match is found
+   *
+   * @param fname
+   * @param notesDict
+   * @param vault If provided, use to filter results
+   * @returns Copy of NoteProps array
+   */
+  static findNotesByFname(
+    fname: string,
+    notesDict: NotePropsFullDict,
+    vault?: DVault
+  ): NoteProps[] {
+    const { notesById, notesByFname } = notesDict;
+    const cleanedFname = cleanName(fname);
+    const ids: string[] | undefined = notesByFname[cleanedFname];
+    if (!ids) {
+      return [];
+    }
+    let notes = ids.map((id) => notesById[id]).filter(isNotUndefined);
+    if (vault) {
+      notes = notes.filter((note) => VaultUtils.isEqualV2(note.vault, vault));
+    }
+    return _.cloneDeep(notes);
+  }
+
+  /**
+   * Remove note from both notesById and notesByFname
+   *
+   * @param note note to delete
+   * @param notesDict
+   */
+  static deleteNote(note: NoteProps, notesDict: NotePropsFullDict) {
+    const { notesById, notesByFname } = notesDict;
+    delete notesById[note.id];
+    NoteFnameDictUtils.deleteNote(note, notesByFname);
+  }
+
+  /**
+   * Add note to notesById and notesByFname.
+   * If note id already exists, check to see if it corresponds to same note by fname.
+   * If fname match, then we only need to update notesById. If fname doesn't match, remove old id from notesByFname first before updating both.
+   *
+   * Otherwise, if note id doesn't exist, add to both dictionaries
+   *
+   * @param note to add
+   * @returns
+   */
+  static addNote(note: NoteProps, notesDict: NotePropsFullDict) {
+    const { notesById, notesByFname } = notesDict;
+    const maybeNote = notesById[note.id];
+    if (maybeNote) {
+      if (cleanName(maybeNote.fname) === cleanName(note.fname)) {
+        notesById[note.id] = note;
+        return;
+      } else {
+        // Remove old fname from fname dict
+        NoteFnameDictUtils.deleteNote(maybeNote, notesByFname);
+      }
+    }
+    notesById[note.id] = note;
+    NoteFnameDictUtils.addNote(note, notesByFname);
+  }
+}
+
+/**
+ * Utilities for working with NotePropsByFnameDict.
+ */
+export class NoteFnameDictUtils {
+  /**
+   * Create a map of {key -> value} where key = noteFname and value is list of ids corresponding to that fname from NotePropsDict
+   *
+   * @param notesById NotePropsDict used to populate map
+   * @returns
+   */
+  static create(notesById: NotePropsDict): NotePropsByFnameDict {
```

Comment:
this might be a big refactor but for consistency, can we rename this to `NotePropsByIdDict`?